Case within Lambda causes java translation error

This code, (with a minimal parser), produces a java file with errors after translation

```
nonterminal Foo;

abstract production bar
f::Foo ::=
{}

function fails
(String ::= Foo) ::=
{
    return (\ f::Foo -> case f of bar() -> "str" end);
}
```

The error for this specific example is 

```
    [javac] /mnt/c/linux_work/silver/generated/src/translateBug/Pfails.java:83: error: incompatible types: <anonymous NodeFactory<Object>> cannot be converted to NodeFactory<StringCatter>
    [javac]             return (common.NodeFactory<common.StringCatter>)((new common.NodeFactory<Object>() {
    [javac]                                                             ^
    [javac] 1 error
```

I don't have much insight into what would be causing it, but I do know some changes that stop the error from happening: 1) If there isn't a case statement, it succeeds, and 2) if `bar()` is replaced with `_`, it succeeds. 

I don't suspect the types involved effect the presence of the error, but this first happened with me in a similar function that operated on `QName`s and `DclInfo`s.

--- a/grammars/silver/modification/lambda_fn/java/Lambda.sv
+++ b/grammars/silver/modification/lambda_fn/java/Lambda.sv
@@ -18,8 +18,8 @@ top::Expr ::= params::ProductionRHS e::Expr
   local finTy :: Type = finalType(top);
   
   top.translation = 
-s"""(new common.NodeFactory<${e.typerep.transType}>() {
-  public final ${e.typerep.transType} invoke(final Object[] args, final Object[] namedArgs) {
+s"""(new common.NodeFactory<${finTy.outputType.transType}>() {
+  public final ${finTy.outputType.transType} invoke(final Object[] args, final Object[] namedArgs) {
     ${params.lambdaTranslation}
     return ${e.translation};
   }

--- a/test/silver_features/Globals.sv
+++ b/test/silver_features/Globals.sv
@@ -28,13 +28,13 @@ equalityTest ( globalbool2, false, Boolean, silver_tests ) ;
 nonterminal Tglob with strGlob;
 synthesized attribute strGlob :: String;
 
-abstract production Tfoo
+abstract production tfoo
 t::Tglob ::=
 {
   t.strGlob = globstring1;
 }
 
-global unT :: Tglob = Tfoo();
+global unT :: Tglob = tfoo();
 
 equalityTest ( unT.strGlob, "Hi", String, silver_tests ) ;
 
@@ -43,7 +43,7 @@ global deT :: Decorated Tglob = decorate unT with {};
 equalityTest ( deT.strGlob, "Hi", String, silver_tests ) ;
 
 wrongCode "initialization expression with type" {
-  global badT :: Decorated Tglob = Tfoo();
+  global badT :: Decorated Tglob = tfoo();
 }
 
 wrongCode "does not have the right signature." { -- TODO: this error message should be improved!!

--- a/test/silver_features/Lambda.sv
+++ b/test/silver_features/Lambda.sv
@@ -34,3 +34,16 @@ global fn::([Integer] ::=) = \ -> [4];
 equalityTest(null(tail(fn())), true, Boolean, silver_tests);
 -- The essential parts are Thunk -> Nodefactory (which captures context) -> Thunk
 
+-- Issue #209 - code gen bug involving casts
+nonterminal LambdaType;
+
+abstract production lambdaType
+f::LambdaType ::=
+{}
+
+function failsLambdaType
+(String ::= LambdaType) ::=
+{
+    return (\ f::LambdaType -> case f of lambdaType() -> "str" end);
+}
+-- End Issue #209

